src/uct/cuda/cuda_ipc/cuda_ipc_md.c

```diff
@@ -106,69 +106,88 @@ uct_cuda_ipc_md_query(uct_md_h md, uct_md_attr_v2_t *md_attr)
 }
 
 static ucs_status_t
-uct_cuda_ipc_mem_reg_push_ctx(CUdeviceptr address, CUdevice *cuda_device_p,
-                              int *is_ctx_pushed)
+uct_cuda_ipc_memh_init(CUdeviceptr address, uct_cuda_ipc_memh_t *memh)
 {
 #define UCT_CUDA_IPC_NUM_ATTRS 2
-    CUdevice cuda_device = CU_DEVICE_INVALID;
     CUpointer_attribute attr_type[UCT_CUDA_IPC_NUM_ATTRS];
     void *attr_data[UCT_CUDA_IPC_NUM_ATTRS];
-    CUcontext cuda_ctx, cuda_curr_ctx;
+    CUcontext cuda_context;
+    CUdevice cuda_device;
     int cuda_device_ordinal;
     ucs_status_t status;
 
-    *is_ctx_pushed = 0;
-
     attr_type[0] = CU_POINTER_ATTRIBUTE_CONTEXT;
-    attr_data[0] = &cuda_ctx;
+    attr_data[0] = &cuda_context;
     attr_type[1] = CU_POINTER_ATTRIBUTE_DEVICE_ORDINAL;
     attr_data[1] = &cuda_device_ordinal;
 
     status = UCT_CUDADRV_FUNC_LOG_ERR(
             cuPointerGetAttributes(UCT_CUDA_IPC_NUM_ATTRS, attr_type, attr_data,
                                    address));
     if (status != UCS_OK) {
-        goto out;
+        return status;
     }
 
-    if (cuda_ctx == NULL) {
-        status = UCT_CUDADRV_FUNC_LOG_ERR(cuDeviceGet(&cuda_device,
-                                                      cuda_device_ordinal));
-        if (status != UCS_OK) {
-            goto out;
-        }
+    ucs_assertv(cuda_device_ordinal >= 0, "cuda_device_ordinal=%d",
+                cuda_device_ordinal);
+
+    status = UCT_CUDADRV_FUNC_LOG_ERR(cuDeviceGet(&cuda_device,
+                                                  cuda_device_ordinal));
+    if (status != UCS_OK) {
+        return status;
+    }
+
+    memh->dev_num  = cuda_device;
+    memh->cuda_ctx = cuda_context;
```

Comment:
why do we need to save both cuda context and cuda device? to avoid extra query of context->device?

Reply:
yes